### PR TITLE
Document rate limiter feature

### PR DESCRIPTION
**Pull request by tanmayv25**
Also fix documentation in test script.

### CODE REVIEW

File: docs/model_configuration.md [hunk 1]
```diff
@@ -501,6 +501,48 @@ the device kind of the instance, for instance, KIND_CPU is "cpu", KIND_MODEL is
   ]
 ```
 
+### Rate Limiter Config
+
+Instance group optionally specify rate limiter config which controls
```
Comment (CoderHam) [inline docs/model_configuration.md:506]: 'rate_limiter' link here itself instead of later

File: docs/model_configuration.md [hunk 2]
```diff
@@ -501,6 +501,48 @@ the device kind of the instance, for instance, KIND_CPU is "cpu", KIND_MODEL is
...
+how [rate limiter](rate_limiter.md) operates on instances in the group.
```
Comment (CoderHam) [inline docs/model_configuration.md:507]: how the rate limiter operates on the instances in the group

File: docs/model_configuration.md [hunk 3]
```diff
@@ -501,6 +501,48 @@ the device kind of the instance, for instance, KIND_CPU is "cpu", KIND_MODEL is
...
+The rate limiter configuration includes the following specification:
+
+#### Resources
+
+The resources required to execute the request on a model instance.
+Resources are just names with corresponding count. They are replicated
```
Comment (CoderHam) [inline docs/model_configuration.md:513]: 'resources' are a mapping of a name for the corresponding resource and their corresponding counts.

File: docs/model_configuration.md [hunk 4]
```diff
@@ -501,6 +501,48 @@ the device kind of the instance, for instance, KIND_CPU is "cpu", KIND_MODEL is
...
+
+```
+  instance_group [
+    {
+      count: 1
+      kind: KIND_GPU
+      rate_limiter {
+        resources [
+          {name: "resource1" count: 4 },
+          {name: "resource2" global: True count: 2 }
```
Comment (CoderHam) [inline docs/model_configuration.md:536]: Use next line for better readability

File: docs/model_configuration.md [hunk 5]
```diff
@@ -501,6 +501,48 @@ the device kind of the instance, for instance, KIND_CPU is "cpu", KIND_MODEL is
...
+
+Instance group optionally specify rate limiter config which controls
+how [rate limiter](rate_limiter.md) operates on instances in the group.
+The rate limiter configuration includes the following specification:
+
+#### Resources
+
+The resources required to execute the request on a model instance.
+Resources are just names with corresponding count. They are replicated
+for each instance. However, resource with `global` flag set `true` will
```
Comment (CoderHam) [inline docs/model_configuration.md:514]: Should this be True instead of true. Not sure it matters

File: docs/model_configuration.md [hunk 6]
```diff
@@ -501,6 +501,48 @@ the device kind of the instance, for instance, KIND_CPU is "cpu", KIND_MODEL is
...
+be shared across the system. Loaded models can not specify a resource
+with the same name both as global and non-global.
+
+#### Priority
+
+The weighting value to be used for prioritizing across instances.
+An instance with priority 2 will be given 1/2 the number of
```
Comment (CoderHam) [inline docs/model_configuration.md:521]: instance_group

File: qa/L0_rate_limiter/test.sh [hunk 7]
```diff
@@ -396,8 +396,9 @@ unset TRITONSERVER_DELAY_SCHEDULER
 ##
 ## Tests with sequence batching
 ##
-# Despite all the possible bs being preferred triton should always form full batches as
-# the second instance would be blocked because of the resource constraints.
+# Send one sequence and check for correct accumulator result. The result should be returned immediately.
+# This test checks whether all the requests are directed to the same instance despite other instances
```
Comment (CoderHam) [inline qa/L0_rate_limiter/test.sh:400]: despite there being other instances ..

**Review by CoderHam — CHANGES_REQUESTED** [review]

File: docs/rate_limiter.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+Rate limiter manages the rate at which requests are scheduled on
+model instances by Triton. The rate limiter operates across all
+models loaded in Triton to allow *cross-model prioritization*.
+
+In absence of rate limiting (--rate-limit=off), Triton schedules
+execution of request (or set of requests when using batching) as
+soon as a model instance is available. This behavior is typically
+best suited for performance. However, there can be cases 
+where running all the models simultaneously place excessive load
```
Comment (Tabrizian) [inline docs/rate_limiter.md:39]: place -> places

File: docs/model_configuration.md [hunk 9]
```diff
@@ -501,6 +501,60 @@ the device kind of the instance, for instance, KIND_CPU is "cpu", KIND_MODEL is
...
+
+Instance group optionally specifies [rate limiter](rate_limiter.md)
+config which controls how the rate limiter operates on the instances
+in the group. The rate limiter configuration includes the following
+specifications:
+
+#### Resources
+
+The set of resources required to execute a model instance. Essentially,
+a resource is just mapping of a name to its corresponding count.
```
Comment (CoderHam) [inline docs/model_configuration.md:514]: "is a mapping" or "is just a mapping"

File: docs/model_configuration.md [hunk 10]
```diff
@@ -501,6 +501,60 @@ the device kind of the instance, for instance, KIND_CPU is "cpu", KIND_MODEL is
...
+By default, the resource counts are local to the device running the
+instance. However, a resource, with `global` flag set to `True`, will
+be shared across the system. Loaded models can not specify a resource
+with the same name both as global and non-global.
+
+#### Priority
+
+Priority specifies the priority of each instance in the group. For an
+instance, priority serves as a weighting value to be used for
```
Comment (CoderHam) [inline docs/model_configuration.md:523]: Isn't the priority set for all instances in that instance group? From the protobuf it appears the priority is paired to an instance group and not a specific instance

**Review by CoderHam — CHANGES_REQUESTED** [review]

File: docs/model_configuration.md (hunk 10 above)
Comment (tanmayv25) [reply docs/model_configuration.md:523]: Yes.. I see why this might become confusing. I will refine the language.

**Review by Tabrizian — APPROVED** [review]

**Review by CoderHam — APPROVED** [review]

File: docs/rate_limiter.md [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+simultaneously may lead to system going out-of-memory.
+
+Rate limiter allows to post-pone the inference execution on some
+model instances such that not all of them runs simultaneously. 
+The model priorities are used to decide which model instance
+to schedule next. 
+
+## Using Rate Limiter
+
+To enable rate limiting users must set `--rate-limit` server
```
Comment (deadeyegoodwin) [inline docs/rate_limiter.md:51]: "option when launching tritonserver"

File: docs/rate_limiter.md [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+The rate limiter is controlled by rate limiter configuration
+described in [rate limiter config](model_configuration.md#rate-limiter-config).
+The rate limiter configuration includes 
+[resources](model_configuration.md#resources) and
+[priority](model_configuration.md#priority) for the instances
+defined by the instance group.
+
+### Resources
+
+Resources are just names with a corresponding count. By default
```
Comment (deadeyegoodwin) [inline docs/rate_limiter.md:64]: "Resources are identified by a unique name and a count indicating the number of copies of the resource. By default, model instances use no...  the model instance indicates that is requires that many..."

When you are talking about the number of resources available I would use "resource copies" and when talking about model instances I would say "resource requirements". "count", "instance", "copy" are used here in multiple ways and that makes it confusing.

Comment (deadeyegoodwin) [reply docs/rate_limiter.md:64]: Also might want to always say "model instance" to be extra clear

File: docs/rate_limiter.md [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+For example, assume three loaded model instances A, B and C each 
+specifying the following resource counts for a single device:
+
+```
+A: [R1: 4, R2: 4]
+B: [R2: 5, R3: 10, R4: 5]
+C: [R1: 1, R3: 7, R4: 2]
+```
+
+By default, the server will select following available resource
```
Comment (deadeyegoodwin) [inline docs/rate_limiter.md:82]: "Based on those model instance requirements, the server will create the following resources with the indicated counts:"

File: docs/rate_limiter.md [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+R3: 10
+R4: 7
+```
+
+These values ensure that all model instances can be successfully
+scheduled. The default for a resource can be overridden by giving
+it explicitly on command-line using `--rate-limit-resource` option.
+`tritonserver --help` will provide with more detailed usage
+instructions.
+
```
Comment (deadeyegoodwin) [inline docs/rate_limiter.md:97]: Here I would say that, by default, resources are per-device and resource requirements for a model instance are enforced against the corresponding resources associated with the device where the instance runs. Then describe how global resources can also be specified in instance group and that global resources are shared across the entire server.

File: docs/model_configuration.md [hunk 15]
```diff
@@ -501,6 +501,59 @@ the device kind of the instance, for instance, KIND_CPU is "cpu", KIND_MODEL is
   ]
 ```
 
+### Rate Limiter Config
+
+Instance group optionally specifies [rate limiter](rate_limiter.md)
+config which controls how the rate limiter operates on the instances
+in the group. The rate limiter configuration includes the following
```
Comment (deadeyegoodwin) [inline docs/model_configuration.md:508]: What if rate limited is enabled and instance doesn't specify any rate limiter config... what is the default?

File: docs/model_configuration.md [hunk 16]
```diff
@@ -501,6 +501,59 @@ the device kind of the instance, for instance, KIND_CPU is "cpu", KIND_MODEL is
...
+specifications:
+
+#### Resources
+
+The set of resources required to execute a model instance. Essentially,
```
Comment (deadeyegoodwin) [inline docs/model_configuration.md:513]: Change "resources" into a link that points to the section in rate limiter section. I don't think you need anything after "Essentially" as you should explain that in rate limiter section (see my comments there). But you should say that the count refers to the number of copies of the resource that the model instance requires to run (see below for other comments on careful use of wording).

File: docs/model_configuration.md [hunk 17]
```diff
@@ -501,6 +501,59 @@ the device kind of the instance, for instance, KIND_CPU is "cpu", KIND_MODEL is
...
+
+#### Priority
+
+Priority serves as a weighting value to be used for prioritizing across
+all the instances of all the models. An instance with priority 2 will be
+given 1/2 the number of scheduling chances as an instance with priority
+1.
+
+The following example specifies the instances in the group requires 
+four "R1" and two "R2" for execution. Resource "R2" is a global
```
Comment (deadeyegoodwin) [inline docs/model_configuration.md:528]: "R2" resources for execution...

File: docs/model_configuration.md [hunk 18]
```diff
@@ -501,6 +501,59 @@ the device kind of the instance, for instance, KIND_CPU is "cpu", KIND_MODEL is
...
+with the same name both as global and non-global.
+
+#### Priority
+
+Priority serves as a weighting value to be used for prioritizing across
+all the instances of all the models. An instance with priority 2 will be
+given 1/2 the number of scheduling chances as an instance with priority
+1.
+
+The following example specifies the instances in the group requires
```
Comment (deadeyegoodwin) [inline docs/model_configuration.md:527]: Say something about how the instance group create 3 model instances (one each on gpus 0 1 and 2). Then you can say explicitly which resources are generates per-device and which are generated globally...

File: docs/model_configuration.md [hunk 19]
```diff
@@ -501,6 +501,59 @@ the device kind of the instance, for instance, KIND_CPU is "cpu", KIND_MODEL is
...
+#### Priority
+
+Priority serves as a weighting value to be used for prioritizing across
+all the instances of all the models. An instance with priority 2 will be
+given 1/2 the number of scheduling chances as an instance with priority
+1.
+
+The following example specifies the instances in the group requires 
+four "R1" and two "R2" for execution. Resource "R2" is a global
+resource. Additionally, the priority of the instance_group is 2.
```
Comment (deadeyegoodwin) [inline docs/model_configuration.md:529]: the rate-limiter priority...

**Review by deadeyegoodwin — CHANGES_REQUESTED** [review]

File: docs/model_configuration.md [hunk 20]
```diff
@@ -501,6 +501,73 @@ the device kind of the instance, for instance, KIND_CPU is "cpu", KIND_MODEL is
   ]
 ```
 
+### Rate Limiter Config
+
+Instance group optionally specifies [rate limiter](rate_limiter.md)
+config which controls how the rate limiter operates on the instances
+in the group. The rate limiter configuration is ignored if rate
+limiting is off. If an instance_group does not provide this
```
Comment (deadeyegoodwin) [inline docs/model_configuration.md:509]: If rate limiting is on and if an instance_group...

File: docs/model_configuration.md [hunk 21]
```diff
@@ -501,6 +501,73 @@ the device kind of the instance, for instance, KIND_CPU is "cpu", KIND_MODEL is
...
+configuration, then the execution on the model instances belonging
+to this group will never be blocked. The configuration includes
```
Comment (deadeyegoodwin) [inline docs/model_configuration.md:511]: ... will not be limited in any way by the rate limiter.

File: docs/rate_limiter.md [hunk 22]
```diff
@@ -0,0 +1,131 @@
...
+In absence of rate limiting (--rate-limit=off), Triton schedules
+execution of a request (or set of requests when using dynamic
+batching) as soon as a model instance is available. This behavior
+is typically best suited for performance. However, there can be
+cases where running all the models simultaneously places excessive
+load on the server. For instance, model execution on some
+frameworks dynamically allocate memory. Running all such models
+simultaneously may lead to system going out-of-memory.
+
+Rate limiter allows to post-pone the inference execution on some
```
Comment (deadeyegoodwin) [inline docs/rate_limiter.md:44]: postpone

File: docs/rate_limiter.md [hunk 23]
```diff
@@ -0,0 +1,131 @@
...
+Resources are identified by a unique name and a count indicating
+the number of copies of the resource. By default, model instance
+uses no rate-limiter resources. By listing a resource/count the
+model instance indicates that it requires that many resources to
+be available on the model instance device before it can be allowed
+to execute. When under execution the specified many resources are
+allocated to the model instance only to be released when the
+execution is over. The available number of resource copies
+are, by default, the max across all model instances that list that
+resource (and so must be adjusted each time a model instance is
```
Comment (deadeyegoodwin) [inline docs/rate_limiter.md:73]: I would remove the parenthetic statement. It makes it seem like the user needs to do something when models are added/removed and it is not clear what they must do

**Review by deadeyegoodwin — CHANGES_REQUESTED** [review]
In the top-level server repo README add some references to rate limiter. I would add a link in the sentence in the user documentation that talks about "scheduling and batching"

**Review by deadeyegoodwin — APPROVED** [review]

**Pull request merged**